source/opt/pass.cpp

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2017 The Khronos Group Inc.
+// Copyright (c) 2017 Valve Corporation
+// Copyright (c) 2017 LunarG Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "pass.h"
+
+#include "iterator.h"
+
+namespace spvtools {
+namespace opt {
+
+namespace {
+
+const uint32_t kEntryPointFunctionIdInIdx = 1;
+
+}  // namespace anonymous
+
+void Pass::AddCalls(ir::Function* func,
+    std::unordered_set<uint32_t>* next) {
+  for (auto bi = func->begin(); bi != func->end(); ++bi)
+    for (auto ii = bi->begin(); ii != bi->end(); ++ii)
+      if (ii->opcode() == SpvOpFunctionCall)
+        next->insert(ii->GetSingleWordInOperand(0));
+}
+
+bool Pass::ProcessEntryPointCallTree(
+    ProcessFunction& pfn, ir::Module* module) {
+  // Map from function's result id to function
+  std::unordered_map<uint32_t, ir::Function*> id2function;
+  id2function.clear();
+  for (auto& fn : *module)
+    id2function[fn.result_id()] = &fn;
+  // Process call tree
+  bool modified = false;
+  std::unordered_set<uint32_t> todo;
```

Comment:
I'm concerned that processing order is unspecified.  This could lead to unstable results across versions or platforms, which would be a pain for testing and debugging.

I think changing "todo" to a std::queue and "next" to a vector would get us most of the way toward deterministic behaviour.  (The entry points iteration is a deterministic scan of a vector.)

Yes there might be duplicated entries in each but they're eliminated by the check against "done".  I'm not concerned about the temporary slight waste of space.  In practice you can reduce some of the space use by filtering against the "done" collection before insertion into "todo";  optional as I think the extra lookup would probably slow it down on average.   But that's just a guess.